File: frontend2/lib/services/message_store.dart
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';
import 'package:sqflite/sqflite.dart';

import '../models/message.dart';

/// Local SQLite cache — mirrors `frontend/src/services/messageStore.ts`.
class MessageStore {
  static Database? _db;

  static Future<void> init() async {
    if (_db != null) return;
    final dir = await getApplicationDocumentsDirectory();
    final path = p.join(dir.path, 'chatapp.db');
    _db = await openDatabase(
      path,
      version: 1,
      onCreate: (db, _) async {
        await db.execute('''
          CREATE TABLE IF NOT EXISTS messages (
            id TEXT PRIMARY KEY,
            server_id INTEGER,
            from_user INTEGER NOT NULL,
            to_user INTEGER NOT NULL,
            type TEXT DEFAULT 'text',
            content TEXT,
            media_url TEXT,
            thumbnail TEXT,
            mime_type TEXT,
            file_name TEXT,
            file_size INTEGER,
            created_at TEXT NOT NULL,
            status TEXT DEFAULT 'sent'
          )
        ''');
        await db.execute(
            'CREATE INDEX IF NOT EXISTS idx_msg_conv ON messages(from_user, to_user, created_at)');
      },
    );
  }

  static Database get _require {
    final db = _db;
    if (db == null) {
      throw StateError('MessageStore not initialised — call init() first.');
    }
    return db;
  }

  static Future<void> saveMessage(Message msg) async {
    await _require.insert(
      'messages',
      msg.toDbRow(),
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
  }

  static Future<List<Message>> getMessages(
    int userId1,
    int userId2, {
    int limit = 30,
    String? beforeId,
  }) async {
    var sql = '''SELECT * FROM messages
      WHERE (from_user = ? AND to_user = ?) OR (from_user = ? AND to_user = ?)''';
    final args = <Object?>[userId1, userId2, userId2, userId1];

    if (beforeId != null) {
      sql +=
          ' AND created_at < (SELECT created_at FROM messages WHERE id = ?)';
      args.add(beforeId);
    }
    sql += ' ORDER BY created_at DESC LIMIT ?';
    args.add(limit);

    final rows = await _require.rawQuery(sql, args);
    return rows.map(Message.fromDbRow).toList();
  }

  static Future<void> updateMessageServerId(
    String localId,
    int serverId,
    String createdAt,
  ) async {
    await _require.update(
      'messages',
      {'server_id': serverId, 'created_at': createdAt, 'status': 'sent'},
      where: 'id = ?',
      whereArgs: [localId],
    );
  }

  static Future<void> updateStatus(String localId, MessageStatus status) async {
    await _require.update(
      'messages',
      {'status': statusToString(status)},
      where: 'id = ?',
      whereArgs: [localId],
    );
  }
}

File: frontend/src/services/messageStore.ts
import * as SQLite from 'expo-sqlite';
import type { Message } from '../types';

let db: SQLite.SQLiteDatabase | null = null;

export async function initDB(): Promise<void> {
  db = await SQLite.openDatabaseAsync('chatapp.db');
  await db.execAsync(`
    CREATE TABLE IF NOT EXISTS messages (
      id TEXT PRIMARY KEY,
      server_id TEXT,
      from_user TEXT NOT NULL,
      to_user TEXT NOT NULL,
      type TEXT DEFAULT 'text',
      content TEXT,
      media_url TEXT,
      local_uri TEXT,
      thumbnail TEXT,
      mime_type TEXT,
      file_name TEXT,
      file_size INTEGER,
      created_at TEXT NOT NULL,
      status TEXT DEFAULT 'sent'
    );
    CREATE INDEX IF NOT EXISTS idx_msg_conv ON messages(from_user, to_user, created_at);
    CREATE TABLE IF NOT EXISTS outbox (
      id TEXT PRIMARY KEY,
      payload TEXT NOT NULL,
      created_at TEXT NOT NULL
    );
  `);
}

function rowToMessage(row: Record<string, unknown>): Message {
  return {
    id: String(row.id),
    serverId: row.server_id ? String(row.server_id) : null,
    from: String(row.from_user),
    to: String(row.to_user),
    type: (row.type as Message['type']) || 'text',
    content: (row.content as string | null) ?? null,
    mediaUrl: (row.media_url as string | null) ?? null,
    localUri: (row.local_uri as string | null) ?? null,
    thumbnail: (row.thumbnail as string | null) ?? null,
    mimeType: (row.mime_type as string | null) ?? null,
    fileName: (row.file_name as string | null) ?? null,
    fileSize: typeof row.file_size === 'number' ? row.file_size : null,
    createdAt: String(row.created_at),
    status: (row.status as Message['status']) || 'sent',
  };
}

export async function saveMessage(msg: Message): Promise<void> {
  if (!db) return;
  await db.runAsync(
    `INSERT OR REPLACE INTO messages
      (id, server_id, from_user, to_user, type, content, media_url, local_uri, thumbnail, mime_type, file_name, file_size, created_at, status)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      msg.id,
      msg.serverId,
      msg.from,
      msg.to,
      msg.type,
      msg.content,
      msg.mediaUrl,
      msg.localUri,
      msg.thumbnail,
      msg.mimeType,
      msg.fileName,
      msg.fileSize,
      msg.createdAt,
      msg.status,
    ],
  );
}

export async function getMessages(userId1: string, userId2: string, limit = 30): Promise<Message[]> {
  if (!db) return [];
  const rows = await db.getAllAsync<Record<string, unknown>>(
    `SELECT * FROM messages
     WHERE (from_user = ? AND to_user = ?) OR (from_user = ? AND to_user = ?)
     ORDER BY created_at DESC
     LIMIT ?`,
    [userId1, userId2, userId2, userId1, limit],
  );
  return rows.map(rowToMessage);
}

export async function updateMessageAck(localId: string, serverId: string, createdAt: string): Promise<void> {
  if (!db) return;
  await db.runAsync(`DELETE FROM messages WHERE id = ? AND id != ?`, [serverId, localId]);
  await db.runAsync(
    `UPDATE messages SET id = ?, server_id = ?, created_at = ?, status = 'sent' WHERE id = ?`,
    [serverId, serverId, createdAt, localId],
  );
}

export async function updateLocalUri(id: string, localUri: string): Promise<void> {
  if (!db) return;
  await db.runAsync(`UPDATE messages SET local_uri = ? WHERE id = ?`, [localUri, id]);
}

export async function enqueueOutbox(id: string, payload: unknown): Promise<void> {
  if (!db) return;
  await db.runAsync(
    `INSERT OR REPLACE INTO outbox (id, payload, created_at) VALUES (?, ?, ?)`,
    [id, JSON.stringify(payload), new Date().toISOString()],
  );
}

export async function getOutbox(): Promise<Array<{ id: string; payload: unknown }>> {
  if (!db) return [];
  const rows = await db.getAllAsync<{ id: string; payload: string }>(`SELECT id, payload FROM outbox ORDER BY created_at ASC`);
  return rows.map((row) => ({ id: row.id, payload: JSON.parse(row.payload) }));
}

export async function removeOutbox(id: string): Promise<void> {
  if (!db) return;
  await db.runAsync(`DELETE FROM outbox WHERE id = ?`, [id]);
}
